posthog/hogql/printer.py

```diff
@@ -1,196 +1,298 @@
                 else:
-                    limit = ast.Call(name="min2", args=[ast.Constant(value=MAX_SELECT_RETURNED_ROWS), limit])
-            elif len(stack) == 1:
-                limit = ast.Constant(value=MAX_SELECT_RETURNED_ROWS)
+                    where = ast.And(exprs=[extra_where, where])
+            else:
+                raise ValueError(f"Invalid where of type {type(extra_where).__name__} returned by join_expr")
+
+            next_join = next_join.next_join
+
+        columns = [self.visit(column) for column in node.select] if node.select else ["1"]
+        where = self.visit(where) if where else None
+        having = self.visit(node.having) if node.having else None
+        prewhere = self.visit(node.prewhere) if node.prewhere else None
+        group_by = [self.visit(column) for column in node.group_by] if node.group_by else None
+        order_by = [self.visit(column) for column in node.order_by] if node.order_by else None
 
         clauses = [
             f"SELECT {'DISTINCT ' if node.distinct else ''}{', '.join(columns)}",
-            f"FROM {from_table}" if from_table else None,
+            f"FROM {' '.join(select_from)}" if len(select_from) > 0 else None,
             "WHERE " + where if where else None,
             f"GROUP BY {', '.join(group_by)}" if group_by and len(group_by) > 0 else None,
             "HAVING " + having if having else None,
             "PREWHERE " + prewhere if prewhere else None,
             f"ORDER BY {', '.join(order_by)}" if order_by and len(order_by) > 0 else None,
         ]
+
+        limit = node.limit
+        if self.context.limit_top_select and len(self.stack) == 1:
+            if limit is not None:
+                if isinstance(limit, ast.Constant) and isinstance(limit.value, int):
+                    limit.value = min(limit.value, MAX_SELECT_RETURNED_ROWS)
+                else:
+                    limit = ast.Call(name="min2", args=[ast.Constant(value=MAX_SELECT_RETURNED_ROWS), limit])
+            else:
+                limit = ast.Constant(value=MAX_SELECT_RETURNED_ROWS)
+
         if limit is not None:
-            clauses.append(f"LIMIT {print_ast(limit, stack, context, dialect)}")
+            clauses.append(f"LIMIT {self.visit(limit)}")
             if node.offset is not None:
-                clauses.append(f"OFFSET {print_ast(node.offset, stack, context, dialect)}")
+                clauses.append(f"OFFSET {self.visit(node.offset)}")
             if node.limit_by is not None:
-                clauses.append(f"BY {', '.join([print_ast(expr, stack, context, dialect) for expr in node.limit_by])}")
+                clauses.append(f"BY {', '.join([self.visit(expr) for expr in node.limit_by])}")
             if node.limit_with_ties:
                 clauses.append("WITH TIES")
 
         response = " ".join([clause for clause in clauses if clause])
-        if len(stack) > 1:
+        if len(self.stack) > 1:
```

Comment:
a comment here would be very nice (I remember what clever stuff this does because of the walkthrough only!)